bootstore/src/db/mod.rs

```diff
@@ -0,0 +1,230 @@
+        &mut self,
+        epoch: i32,
+        share: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        use schema::key_share_prepares::dsl;
+        // We save the digest so we don't have to deserialize and recompute most of the time.
+        // We'd only want to do that for a consistency check occasionally.
+        let val = serde_json::to_string(&share)?;
+        let share_digest =
+            sprockets_common::Sha3_256Digest(Sha3_256::digest(&val).into())
+                .into();
+        let prepare =
+            KeySharePrepare { epoch, share: Share(share), share_digest };
+        diesel::insert_into(dsl::key_share_prepares)
+            .values(&prepare)
+            .execute(&mut self.conn)?;
+        Ok(())
+    }
+
+    pub fn commit_share(
+        &mut self,
+        epoch: i32,
+        digest: sprockets_common::Sha3_256Digest,
+    ) -> Result<(), Error> {
+        use schema::key_share_commits;
+        use schema::key_share_prepares;
+        let commit =
+            KeyShareCommit { epoch, share_digest: digest.clone().into() };
+        self.conn.immediate_transaction(|tx| {
+            // We only want to commit if the share digest of the commit is the
+            // same as that of the prepare.
+            let prepare_digest = key_share_prepares::table
+                .select(key_share_prepares::share_digest)
+                .filter(key_share_prepares::epoch.eq(epoch))
+                .get_result::<Sha3_256Digest>(tx)?;
+
+            if prepare_digest != digest.into() {
+                return Err(Error::CommitHashMismatch { epoch });
+            }
+
+            diesel::insert_into(key_share_commits::table)
+                .values(&commit)
+                .execute(tx)?;
+            Ok(())
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::trust_quorum::{RackSecret, ShareDistribution};
+    use omicron_test_utils::dev::test_setup_log;
+    use rand::distributions::Alphanumeric;
+    use rand::{thread_rng, Rng};
+
+    // TODO: Fill in with actual member certs
+    fn new_shares() -> Vec<ShareDistribution> {
+        let member_device_id_certs = vec![];
+        let rack_secret_threshold = 3;
+        let total_shares = 5;
+        let secret = RackSecret::new();
+        let (shares, verifier) =
+            secret.split(rack_secret_threshold, total_shares).unwrap();
+
+        shares
+            .into_iter()
+            .map(move |share| ShareDistribution {
+                threshold: rack_secret_threshold,
+                verifier: verifier.clone(),
+                share,
+                member_device_id_certs: member_device_id_certs.clone(),
+            })
+            .collect()
+    }
+
+    fn rand_db_name() -> String {
+        let seed: String = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+        format!("/tmp/testdb-{}.sqlite", seed)
```

Comment:
Could we use `:memory:` instead for tests, to not have to worry about temp files? If not, do we need to wrap this is something that will delete it when we're done? (Maybe using `tempfile`?)

Reply:
I specifically didn't worry about cleanup because:
1. These DBs are going to be quite small.
2. We probably want to be able to inspect them on test failures.

1 is just a lazy punt on my part and not a justification. 2 is potentially important, but maybe we can hook into cargo test somehow and output a file only on failure?

In any case, for now I will convert to :memory: as you suggest. That makes the most sense in the short term.

Reply:
I think there are checks in CI that successful tests do not leave leftovers behind in `/tmp`. If you want to keep them around for failures, the simplest thing might be to use something like https://docs.rs/tempfile/latest/tempfile/struct.Builder.html#method.tempdir to create a temp directory that will be deleted on drop, and put the file in there.

Reply:
Fixed to use `:memory:` in https://github.com/oxidecomputer/omicron/pull/1656/commits/a9ac17c3df90c34267b77f05f6b1a1730658d76c